File: libraries/ios/Sources/VoidhashCore/Caching/CacheManager.swift
import Foundation

/// JSON envelope persisted for every cached value.
///
/// The shape is a cross-platform contract: `{value, createdAt, staleAt, expiresAt}` with
/// millisecond epoch timestamps, matching `src/core/caching/cache-manager.ts`.
struct CacheEnvelope<Value: Codable & Sendable>: Codable, Sendable {
    let value: Value
    let expiresAt: Double?
    let createdAt: Double
    let staleAt: Double?

    private enum CodingKeys: String, CodingKey {
        case createdAt
        case expiresAt
        case staleAt
        case value
    }

    init(value: Value, createdAt: Double, expiresAt: Double?, staleAt: Double?) {
        self.value = value
        self.createdAt = createdAt
        self.expiresAt = expiresAt
        self.staleAt = staleAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        value = try container.decode(Value.self, forKey: .value)
        createdAt = try container.decode(Double.self, forKey: .createdAt)
        expiresAt = try container.decodeIfPresent(Double.self, forKey: .expiresAt)
        staleAt = try container.decodeIfPresent(Double.self, forKey: .staleAt)
    }

    // The TypeScript implementation writes explicit `null`s rather than omitting the keys;
    // encodeIfPresent would drop them and break byte-level parity of the persisted envelope.
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(createdAt, forKey: .createdAt)
        if let expiresAt {
            try container.encode(expiresAt, forKey: .expiresAt)
        } else {
            try container.encodeNil(forKey: .expiresAt)
        }
        if let staleAt {
            try container.encode(staleAt, forKey: .staleAt)
        } else {
            try container.encodeNil(forKey: .staleAt)
        }
        try container.encode(value, forKey: .value)
    }
}

/// A cache entry plus its freshness flags.
public struct CacheHit<Value: Sendable>: Sendable {
    /// The decoded cached value.
    public let value: Value
    /// Millisecond epoch timestamp of when the entry was written.
    public let createdAt: Double
    /// Millisecond epoch timestamp after which the entry is considered stale, if any.
    public let staleAt: Double?
    /// Millisecond epoch timestamp after which the entry is considered expired, if any.
    public let expiresAt: Double?
    /// Whether `staleAt` or `expiresAt` has elapsed.
    public let isStale: Bool
    /// Whether `expiresAt` has elapsed. Expired entries are still returned: TTL drives how
    /// urgently the SDK refreshes, never whether a value is available offline.
    public let isExpired: Bool
}

/// A value paired with the freshness of the cache entry it came from.
///
/// Returned by the `…State` read accessors so callers can decide how much to trust a value that
/// was served offline, without changing the signature of the plain accessors.
public struct Stale<Value: Sendable>: Sendable {
    /// The value, cached or freshly fetched.
    public let value: Value
    /// Whether the cached entry had passed its stale deadline.
    public let isStale: Bool
    /// Whether the cached entry had passed its expiry deadline.
    public let isExpired: Bool

    public init(value: Value, isStale: Bool = false, isExpired: Bool = false) {
        self.value = value
        self.isStale = isStale
        self.isExpired = isExpired
    }

    /// Maps the wrapped value, keeping the freshness flags.
    public func map<Mapped: Sendable>(_ transform: (Value) -> Mapped) -> Stale<Mapped> {
        return Stale<Mapped>(
            value: transform(value), isStale: isStale, isExpired: isExpired)
    }
}

/// Envelope-based cache with a `cache-keys` index enabling ``clear()``.
///
/// Expired reads do not delete: an expired entry comes back with `isExpired == true` so an offline
/// device keeps serving the last known state. Entries are removed by an overwrite, an explicit
/// delete/clear, or when a corrupt envelope is discarded. Writes register the key in a
/// deduplicated index.
public actor CacheManager {
    /// Index key holding the JSON array of every key written through this manager.
    public static let cacheKeysKey = "cache-keys"

    private let adapter: CacheAdapter
    private let now: @Sendable () -> Double
    private let diagnostics: DiagnosticEmitter
    /// The `cache-keys` index, read from the adapter once and mutated in memory from then on.
    ///
    /// Every adapter call suspends the actor, so two writers reading the index back on each
    /// write could interleave: both would read the same list, and the second write would drop
    /// the first's key. An entry missing from the index is invisible to `clear` and
    /// `deleteByPrefix`, so it would outlive the identity it belongs to.
    private var indexedKeys: [String]?

    /// - Parameters:
    ///   - adapter: Backing string storage.
    ///   - now: Millisecond epoch clock, injectable for tests.
    ///   - diagnostics: Receives `CACHE_READ_FAILED` for corrupt entries.
    public init(
        adapter: CacheAdapter,
        now: @escaping @Sendable () -> Double = { Date().timeIntervalSince1970 * 1000 },
        diagnostics: DiagnosticEmitter = DiagnosticEmitter(nil)
    ) {
        self.adapter = adapter
        self.now = now
        self.diagnostics = diagnostics
    }

    /// Reads and decodes a cached value.
    ///
    /// An expired entry is returned with `isExpired == true` rather than deleted; a corrupt one
    /// is deleted and reported as a miss with a `CACHE_READ_FAILED` diagnostic.
    public func get<Value: Codable & Sendable>(_ key: String, as type: Value.Type = Value.self)
        async -> CacheHit<Value>?
    {
        guard let raw = await adapter.get(key), let data = raw.data(using: .utf8) else {
            return nil
        }
        guard let envelope = try? JSONDecoder().decode(CacheEnvelope<Value>.self, from: data) else {
            diagnostics.emit(
                .cache, code: "CACHE_READ_FAILED", operation: "cache.get",
                message: "Discarding the undecodable cache entry \"\(key)\"")
            await delete(key)
            return nil
        }

        // A `0` timestamp is falsy in the TypeScript implementation, so it reads as "never" here
        // too rather than as "expired at the epoch".
        let timestamp = now()
        let isExpired = CacheManager.hasElapsed(envelope.expiresAt, at: timestamp)
        let isStale = isExpired || CacheManager.hasElapsed(envelope.staleAt, at: timestamp)

        return CacheHit(
            value: envelope.value,
            createdAt: envelope.createdAt,
            staleAt: envelope.staleAt,
            expiresAt: envelope.expiresAt,
            isStale: isStale,
            isExpired: isExpired
        )
    }

    /// Writes a value and registers its key in the `cache-keys` index.
    ///
    /// - Parameters:
    ///   - ttl: Lifetime in milliseconds after which the entry is flagged expired and should be
    ///     refreshed. The value remains available offline. `nil` or `0` writes a `null` deadline:
    ///     the entry never expires.
    ///   - staleTime: Milliseconds after which the entry is still served but flagged stale. `nil`
    ///     or `0` writes a `null` deadline: the entry is never stale.
    public func set<Value: Codable & Sendable>(
        _ key: String,
        value: Value,
        ttl: Double? = nil,
        staleTime: Double? = nil
    ) async {
        let timestamp = now()
        let envelope = CacheEnvelope(
            value: value,
            createdAt: timestamp,
            expiresAt: CacheManager.deadline(after: ttl, at: timestamp),
            staleAt: CacheManager.deadline(after: staleTime, at: timestamp)
        )

        guard let data = try? JSONEncoder().encode(envelope),
            let json = String(data: data, encoding: .utf8)
        else {
            return
        }

        await adapter.set(key, value: json)
        await storeCacheKey(key)
    }

    /// Removes a single entry and drops it from the key index.
    public func delete(_ key: String) async {
        await adapter.delete(key)
        guard key != CacheManager.cacheKeysKey else {
            return
        }
        let keys = await getCacheKeys()
        if keys.contains(key) {
            await writeCacheKeys(keys.filter { $0 != key })
        }
    }

    /// Deletes every indexed key and the index itself.
    public func clear() async {
        let keys = await getCacheKeys()
        indexedKeys = []
        for key in keys {
            await adapter.delete(key)
        }
        await adapter.delete(CacheManager.cacheKeysKey)
    }

    /// Returns the keys registered in the `cache-keys` index, deduplicated in write order.
    public func getCacheKeys() async -> [String] {
        if let indexedKeys {
            return indexedKeys
        }
        let loaded = await loadCacheKeys()
        // A write that landed while the read was suspended already owns the index.
        if let indexedKeys {
            return indexedKeys
        }
        indexedKeys = loaded
        return loaded
    }

    /// Deletes every key whose name starts with `prefix` and drops them from the index.
    ///
    /// Used by `identify` and `reset` to invalidate the per-person entries of the identity being
    /// left behind.
    public func deleteByPrefix(_ prefix: String) async {
        let keys = await getCacheKeys()
        let matched = keys.filter { $0.hasPrefix(prefix) }
        guard !matched.isEmpty else {
            return
        }
        await writeCacheKeys(keys.filter { !$0.hasPrefix(prefix) })
        for key in matched {
            await adapter.delete(key)
        }
    }

    private func loadCacheKeys() async -> [String] {
        guard let raw = await adapter.get(CacheManager.cacheKeysKey),
            let data = raw.data(using: .utf8),
            let keys = try? JSONDecoder().decode([String].self, from: data)
        else {
            return []
        }
        var seen: Set<String> = []
        return keys.filter { seen.insert($0).inserted }
    }

    // `options?.ttl ? Date.now() + options.ttl : null` in the TypeScript SDK: a zero duration is
    // falsy there and means "no deadline", not "already elapsed".
    private static func deadline(after duration: Double?, at timestamp: Double) -> Double? {
        guard let duration, duration != 0 else {
            return nil
        }
        return timestamp + duration
    }

    private static func hasElapsed(_ deadline: Double?, at timestamp: Double) -> Bool {
        guard let deadline, deadline != 0 else {
            return false
        }
        return deadline < timestamp
    }

    // The index is a set: keys rewritten on every capture (the analytics session) must not grow
    // it with each write. Reads compact any duplicates a previous SDK version left behind.
    private func storeCacheKey(_ key: String) async {
        var keys = await getCacheKeys()
        guard !keys.contains(key) else {
            return
        }
        keys.append(key)
        await writeCacheKeys(keys)
    }

    // The in-memory index is updated before the write suspends, so a concurrent writer that
    // reads it next already sees this change and its own write carries both.
    private func writeCacheKeys(_ keys: [String]) async {
        indexedKeys = keys
        guard let data = try? JSONEncoder().encode(keys),
            let json = String(data: data, encoding: .utf8)
        else {
            return
        }
        await adapter.set(CacheManager.cacheKeysKey, value: json)
    }
}

File: libraries/android/sdk/src/main/java/com/voidhash/sdk/cache/CacheManager.kt
package com.voidhash.sdk.cache

import com.voidhash.sdk.diagnostics.DiagnosticEmitter
import com.voidhash.sdk.diagnostics.VoidhashDiagnosticKind
import com.voidhash.sdk.storage.PersistenceWriter
import kotlinx.coroutines.CompletableDeferred
import org.json.JSONArray
import org.json.JSONObject
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch

private const val CACHE_KEYS_KEY = "cache-keys"

/** Key/value string store backing [CacheManager]. */
interface CacheAdapter {
    /** Returns the stored envelope for [key], or `null` when nothing is stored. */
    fun get(key: String): String?

    /** Stores [value] under [key], replacing anything already there. */
    fun set(key: String, value: String)

    /** Removes [key]; a missing key is not an error. */
    fun delete(key: String)

    /**
     * Every entry this adapter owns, or `null` when it cannot enumerate itself.
     *
     * An adapter that can answer this lets [CacheManager] load the whole namespace once, on
     * the writer thread, after which no read ever touches storage again.
     */
    fun snapshot(): Map<String, String>? = null
}

/** In-memory adapter, useful for tests and for `enabled = false` clients. */
class InMemoryCacheAdapter : CacheAdapter {
    private val values = mutableMapOf<String, String>()

    override fun get(key: String): String? = synchronized(values) { values[key] }

    override fun set(key: String, value: String) {
        synchronized(values) { values[key] = value }
    }

    override fun delete(key: String) {
        synchronized(values) { values.remove(key) }
    }

    override fun snapshot(): Map<String, String> = synchronized(values) { values.toMap() }
}

/** A cached entry plus its freshness verdicts. */
data class CacheHit<T>(
    val value: T,
    val createdAt: Long,
    val staleAt: Long?,
    val expiresAt: Long?,
    val isStale: Boolean,
    val isExpired: Boolean,
)

/**
 * JSON-envelope cache shared with the other Voidhash SDKs.
 *
 * Every entry is stored as `{value, createdAt, staleAt, expiresAt}`. An expired entry is
 * still returned, flagged `isExpired`: the TTL says how urgently the value should be
 * refreshed, not whether it may be used. An offline device that serves a two-day-old
 * entitlement is doing the right thing; one that reports "no entitlement" because the TTL
 * lapsed is not. Entries are removed only by an overwrite, [delete], or [clear].
 *
 * Written keys are tracked in a deduplicated `cache-keys` index so [clear] can drop the
 * whole namespace.
 */
class CacheManager(
    private val adapter: CacheAdapter,
    private val diagnostics: DiagnosticEmitter = DiagnosticEmitter(),
    /**
     * Serializes adapter writes off the caller's thread and performs the one-time warm-up.
     * Without one the cache reads and writes inline, which is only appropriate in tests.
     */
    private val writer: PersistenceWriter? = null,
    /** Runs on the writer thread before the first warm read; used for the legacy migration. */
    private val onWarmUp: () -> Unit = {},
    // Last so `CacheManager(adapter) { now }` still binds the trailing lambda to the clock.
    private val clock: () -> Long = { System.currentTimeMillis() },
) {
    /**
     * Write-through mirror of the adapter.
     *
     * Reads are answered from here so no caller waits on `SharedPreferences`, which parses
     * its whole backing file the first time it is touched. Before the warm-up a key that is
     * not mirrored yet falls through to the adapter; after it a memory miss is the answer.
     */
    private val memory = ConcurrentHashMap<String, String>()

    /** Keys erased before the warm-up, which the snapshot must not resurrect. */
    // `ConcurrentHashMap.newKeySet` is API 24; this is the same set on API 23.
    private val absent: MutableSet<String> = Collections.newSetFromMap(ConcurrentHashMap())
    private val cacheKeysLock = Any()
    private val warmed = CompletableDeferred<Unit>()
    private val warmedLatch = CountDownLatch(1)

    /** Set once the whole namespace is mirrored; from then on reads never touch storage. */
    @Volatile
    private var fullyLoaded = false

    /** Whether the one-time storage snapshot and migration have completed. */
    val isWarm: Boolean get() = warmed.isCompleted

    init {
        if (writer == null) warmUp() else writer.submit(::warmUp)
    }

    /**
     * Mirrors the namespace into memory once, on the writer thread.
     *
     * Anything written before this lands — the host's `distinctId` option, an identity
     * minted by an early capture — is newer than what storage holds, and its own write is
     * already queued behind this task. The snapshot therefore only fills gaps: a key already
     * in memory keeps its value, and a key erased before the warm-up stays erased.
     */
    private fun warmUp() {
        try {
            onWarmUp()
            adapter.snapshot()?.let { stored ->
                for ((key, value) in stored) {
                    if (!absent.contains(key)) memory.putIfAbsent(key, value)
                }
                fullyLoaded = true
            }
            compactCacheKeys()
        } finally {
            warmed.complete(Unit)
            warmedLatch.countDown()
        }
    }

    /**
     * Suspends until the one-time warm-up — legacy migration and key-index compaction — has
     * run. Callers that need a correct identity (analytics start-up, the first schema
     * resolve) await this; interactive reads do not, and fall through to the adapter.
     */
    suspend fun awaitWarm() {
        warmed.await()
    }

    /**
     * Blocks the calling thread until the warm-up has run.
     *
     * For the rare non-suspending caller that must not act on a miss it cannot trust: an
     * identity minted before the legacy migration ran would split the device into two
     * people. Returns immediately once warm, so it costs nothing after start-up.
     */
    fun awaitWarmBlocking() {
        warmedLatch.await()
    }

    private fun read(key: String): String? {
        memory[key]?.let { return it }
        // Once the namespace is mirrored, a memory miss *is* the answer.
        if (fullyLoaded || absent.contains(key)) return null
        // Only reachable before the warm-up lands, which is before `initialize` returns. A
        // miss here is not remembered: the legacy migration may still produce the key.
        val stored = adapter.get(key)
        if (stored != null) memory.putIfAbsent(key, stored)
        return memory[key]
    }

    private fun write(key: String, value: String) {
        memory[key] = value
        absent.remove(key)
        if (writer == null) adapter.set(key, value) else writer.submit { adapter.set(key, value) }
    }

    private fun erase(key: String) {
        memory.remove(key)
        absent.add(key)
        if (writer == null) adapter.delete(key) else writer.submit { adapter.delete(key) }
    }

    /** Reads the raw envelope for [key]. */
    fun get(key: String): CacheHit<Any?>? {
        val stored = read(key) ?: return null
        val envelope = runCatching { JSONObject(stored) }.getOrNull()
        val createdAt = (envelope?.opt("createdAt") as? Number)?.toLong()
        val hasValue = envelope?.has("value") == true
        val validDeadlines = envelope != null &&
            listOf("staleAt", "expiresAt").all { deadline ->
                !envelope.has(deadline) || envelope.isNull(deadline) || envelope.opt(deadline) is Number
            }
        if (envelope == null || createdAt == null || !hasValue || !validDeadlines) {
            // A corrupt entry is a miss, never a failure: the caller refetches.
            diagnostics.emit(
                VoidhashDiagnosticKind.CACHE,
                code = "CACHE_READ_FAILED",
                operation = "cache.get",
                message = "Discarded an unreadable cache entry for $key",
            )
            delete(key)
            return null
        }

        val now = clock()
        val expiresAt = envelope.deadlineOrNull("expiresAt")
        val staleAt = envelope.deadlineOrNull("staleAt")
        val isExpired = expiresAt != null && expiresAt < now

        return CacheHit(
            value = envelope.opt("value").takeUnless { it === JSONObject.NULL },
            createdAt = createdAt,
            staleAt = staleAt,
            expiresAt = expiresAt,
            // An expired entry is stale by definition, even when no stale window was set.
            isStale = isExpired || (staleAt != null && staleAt < now),
            isExpired = isExpired,
        )
    }

    /** Reads [key] as a string entry. */
    fun getString(key: String): CacheHit<String>? = getTyped(key)

    /** Reads [key] as an object entry. */
    fun getObject(key: String): CacheHit<JSONObject>? = getTyped(key)

    /** Reads [key] as an array entry. */
    fun getArray(key: String): CacheHit<JSONArray>? = getTyped(key)

    @PublishedApi
    internal fun reportTypeMismatch(key: String, expected: String) {
        diagnostics.emit(
            VoidhashDiagnosticKind.CACHE,
            code = "CACHE_READ_FAILED",
            operation = "cache.get",
            message = "The cache entry for $key is not a $expected; treating it as a miss",
        )
    }

    private inline fun <reified T : Any> getTyped(key: String): CacheHit<T>? {
        val hit = get(key) ?: return null
        val value = hit.value as? T
        if (value == null) {
            // A value of the wrong shape is a miss, but a silent one would hide a real
            // encoding bug behind an endless stream of refetches.
            reportTypeMismatch(key, T::class.java.simpleName)
            return null
        }
        return CacheHit(
            value = value,
            createdAt = hit.createdAt,
            staleAt = hit.staleAt,
            expiresAt = hit.expiresAt,
            isStale = hit.isStale,
            isExpired = hit.isExpired,
        )
    }

    /**
     * Writes [value] under [key], optionally with a time-to-live and a stale-after window
     * (both in milliseconds). A `null` or `0` window means "never expires" / "never stale"
     * and is written as an explicit `null`, matching the envelope the other Voidhash SDKs
     * write.
     */
    fun set(key: String, value: Any?, ttlMs: Long? = null, staleTimeMs: Long? = null) {
        val now = clock()
        val envelope = JSONObject().apply {
            put("createdAt", now)
            put("expiresAt", ttlMs?.takeIf { it != 0L }?.let { now + it } ?: JSONObject.NULL)
            put("staleAt", staleTimeMs?.takeIf { it != 0L }?.let { now + it } ?: JSONObject.NULL)
            put("value", value ?: JSONObject.NULL)
        }
        write(key, envelope.toString())
        storeCacheKey(key)
    }

    /** Removes a single entry. */
    fun delete(key: String) {
        synchronized(cacheKeysLock) {
            erase(key)
            if (key == CACHE_KEYS_KEY) return
            val keys = readCacheKeys()
            if (keys.remove(key)) writeCacheKeys(keys)
        }
    }

    /** Removes every entry whose key starts with [prefix]. */
    fun deleteByPrefix(prefix: String) {
        synchronized(cacheKeysLock) {
            val remaining = linkedSetOf<String>()
            for (key in readCacheKeys()) {
                if (key.startsWith(prefix)) erase(key) else remaining.add(key)
            }
            writeCacheKeys(remaining)
        }
    }

    /** Removes every entry recorded in the `cache-keys` index. */
    fun clear() {
        synchronized(cacheKeysLock) {
            for (key in readCacheKeys()) {
                erase(key)
            }
            erase(CACHE_KEYS_KEY)
        }
    }

    /** Returns the tracked cache keys, in insertion order and without duplicates. */
    fun getCacheKeys(): List<String> = readCacheKeys().toList()

    private fun readCacheKeys(): LinkedHashSet<String> {
        val stored = read(CACHE_KEYS_KEY) ?: return linkedSetOf()
        val keys = runCatching { JSONArray(stored) }.getOrNull() ?: return linkedSetOf()
        val unique = linkedSetOf<String>()
        for (index in 0 until keys.length()) {
            keys.optString(index).takeIf { it.isNotEmpty() }?.let(unique::add)
        }
        return unique
    }

    /**
     * Adds [key] to the index.
     *
     * Read-modify-write under a lock: two concurrent writers reading the same index and
     * writing back their own copy would lose one of the keys, and a lost key is an entry
     * [clear] can never remove.
     */
    private fun storeCacheKey(key: String) {
        synchronized(cacheKeysLock) {
            val keys = readCacheKeys()
            if (!keys.add(key)) return
            writeCacheKeys(keys)
        }
    }

    private fun writeCacheKeys(keys: Set<String>) {
        write(CACHE_KEYS_KEY, JSONArray(keys.toList()).toString())
    }

    /**
     * Rewrites the index once at construction, dropping the duplicates earlier SDK versions
     * appended on every write. Left alone, that index grew without bound and made [clear]
     * O(writes) instead of O(keys).
     */
    private fun compactCacheKeys() {
        synchronized(cacheKeysLock) {
            val stored = read(CACHE_KEYS_KEY) ?: return
            val unique = readCacheKeys()
            val compacted = JSONArray(unique.toList()).toString()
            if (compacted != stored) {
                write(CACHE_KEYS_KEY, compacted)
            }
        }
    }

    /**
     * Reads a deadline field. A real deadline is always an epoch-millis timestamp, so `0` is
     * the shared "no deadline" sentinel and must not read as a moment long past.
     */
    private fun JSONObject.deadlineOrNull(key: String): Long? {
        if (!has(key) || isNull(key)) return null
        return optLong(key).takeUnless { it == 0L }
    }
}
